### ISSUE TITLE
GLocationService Crash

### ISSUE THREAD

**Issue by jruston** [GrapheneOS/os-issue-tracker#4603]
I am seeing a number of crashes coming from GLocationService in my third-party app. The crashes look like the below log. So far all reports have been when the app was in the background, and the devices register in Crashlytics as "Android 15". It seems to be specific to GrapheneOS however, as it does not occur on devices which aren't running GrapheneOS. I have been unable to find which build the users have.

          Fatal Exception: com.google.android.gms.common.api.ApiException: 15: 
       at com.google.android.gms.common.internal.ApiExceptionUtil.fromStatus(com.google.android.gms:play-services-base@@18.4.0:3)
       at com.google.android.gms.common.api.internal.TaskUtil.setResultOrApiException(com.google.android.gms:play-services-base@@18.4.0:4)
       at com.google.android.gms.internal.location.zzdk.zzb(com.google.android.gms:play-services-location@@21.2.0:1)
       at com.google.android.gms.internal.location.zzy.zza(com.google.android.gms:play-services-location@@21.2.0:4)
       at com.google.android.gms.internal.location.zzb.onTransact(com.google.android.gms:play-services-location@@21.2.0:3)
       at android.os.Binder.transact(Binder.java:1255)
       at com.google.android.gms.location.internal.ILocationStatusCallback$Stub$Proxy.onResult(ILocationStatusCallback.java:100)
       at app.grapheneos.gmscompat.location.GLocationService$getCurrentLocation$consumer$1.accept(GLocationService.kt:232)
       at app.grapheneos.gmscompat.location.GLocationService$getCurrentLocation$consumer$1.accept(GLocationService.kt:218)
       at android.location.LocationManager$GetCurrentLocationTransport$1.operate(LocationManager.java:3286)
       at android.location.LocationManager$GetCurrentLocationTransport$1.operate(LocationManager.java:3283)
       at com.android.internal.listeners.ListenerExecutor.lambda$executeSafely$0(ListenerExecutor.java:127)
       at com.android.internal.listeners.ListenerExecutor$$ExternalSyntheticLambda0.run(D8$$SyntheticClass)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
       at java.lang.Thread.run(Thread.java:1117)

**Comment by dschotman** [comment]
Is this why my location hasn't been working since the last 2024121200 update on my Pixel 7 Pro? Or is it a separate issue?

Opening Google Maps won't load the current location, only after I navigate somewhere and exit the navgation. Google Maps Timeline stays empty. Polarsteps is not updating location anymore. Dating apps cannot read current location anymore. And the green location indicator in the upper right corner hasn't been showing meaning the location hasn't been requested. 

I've been trying quite a few things like rebooting and changing settings, but nothing seems to work. It is especially frustrating I cannot use the travel tracker within Polarsteps anymore because it simply doesn't work without background location updates.

Will updating to the latest beta help? Please let me know if I can assist.

**Comment by Cyber-Typhoon** [comment]
![Screenshot_20250103-000905](https://github.com/user-attachments/assets/e26ba9a0-602a-414e-b94e-5bf479fe1faf)
> disabled location rerouting

Are you talking about the setting in the screenshot attached? 

I have it like in the screenshot and Google Maps and Osmand are both not getting my location anymore.

**Comment by Cyber-Typhoon** [comment]
> @Cyber-Typhoon Are you indoors without GNSS satellite reception? You don't have it configured to network-based location so and it's likely working as intended. Use the GPSTest app.

Thank you. You are right. It is not related, I installed the GPSTest app and moved to a zone with 3 or more GNSS satellite and it got my location. Sorry for convoluting the discussion. 

![Screenshot_20250103-002601](https://github.com/user-attachments/assets/732ee875-9f59-4cb3-acbf-527da2436cf5)

**Comment by dschotman** [comment]
> Cleaned up the unrelated issues which have been resolved.

I don't agree with that. Even after updating to build `2024123000`, I still have the same issues on my Pixel 7 Pro, nothing with background location does work.
Can you point me in the direction where this has been resolved?

**Comment by jruston** [comment]
@thestinger The crash seems to only affect apps which are using location in the background. 100% of crashes occur when the app is in the background. For the affected users, they see this crash extremely frequently. It is possible that background location is not working correctly at all for these users. Sadly I have no information about their device configuration. I don't think most affected users are aware this is happening, but it is stopping the app from working correctly.

The code 15 that Play services returns in the exception seems to indicate that the request to Play services timed out, and that the timeout is not being handled by GrapheneOS. It is unclear why it would continuously time out for these users when the app is in the background, but in any case GrapheneOS should handle timeout errors in order to prevent the whole app from crashing every time.
